nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+// can be used via the returned handle to interact with the operation.
+//
+// Use the returned [OperationHandle] to retrieve the operation's result.
+func (c *Client) StartOperation(ctx context.Context, request StartOperationRequest) (*OperationHandle, error) {
+	if closer, ok := request.Body.(io.Closer); ok {
+		// Close the request body in case we error before sending the HTTP request.
+		defer closer.Close()
+	}
+	url, err := c.joinURL(request.Operation)
+	if err != nil {
+		return nil, err
+	}
+	if request.CallbackURL != "" {
+		q := url.Query()
+		q.Set(nexusapi.QueryCallbackURL, request.CallbackURL)
+		url.RawQuery = q.Encode()
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", url.String(), request.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if request.Header != nil {
+		httpReq.Header = request.Header.Clone()
+	}
+	if request.RequestID == "" {
+		requestIDFromHeader := request.Header.Get(nexusapi.HeaderRequestID)
+		if requestIDFromHeader != "" {
+			request.RequestID = requestIDFromHeader
+		} else {
+			request.RequestID = uuid.NewString()
+		}
+	}
+	httpReq.Header.Set(nexusapi.HeaderRequestID, request.RequestID)
+
+	httpReq.Header.Set(headerUserAgent, UserAgent)
+	response, err := c.Options.HTTPClient.Do(httpReq)
+	if err != nil {
+		return nil, err
+	}
+	// Do not close response body here to allow successful result to read it.
+	switch response.StatusCode {
+	case 200, 204:
+		handle := &OperationHandle{
+			operation: request.Operation,
+			state:     nexusapi.OperationStateSucceeded,
+			response:  response,
+		}
+		return handle, nil
+	}
+
+	// Do this once here and make sure it doesn't leak.
+	defer response.Body.Close()
+
+	switch response.StatusCode {
+	case 201:
+		info, err := c.operationInfoFromResponse(response)
+		if err != nil {
+			return nil, err
+		}
+		if info.State != nexusapi.OperationStateRunning {
+			return nil, c.newUnexpectedResponseError(fmt.Sprintf("invalid operation state in response info: %q", info.State), response)
+		}
+		handle := &OperationHandle{
+			client:    c,
+			operation: request.Operation,
+			id:        info.ID,
+			state:     info.State,
+			response:  response,
+		}
+		return handle, nil
+	case nexusapi.StatusOperationFailed:
+		state, err := c.getUnsuccessfulStateFromHeader(response)
+		if err != nil {
+			return nil, err
+		}
+
+		failure, err := c.failureFromResponse(response)
+		if err != nil {
+			return nil, err
+		}
+
+		return &OperationHandle{
+			operation: request.Operation,
+			state:     state,
+			failure:   failure,
+		}, nil
+	default:
+		return nil, c.newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response)
+	}
+}
+
+type getOperationResultRequest struct {
+	Operation   string
+	OperationID string
+	Wait        bool
+	Header      http.Header
+}
+
+func (c *Client) operationInfoFromResponse(response *http.Response) (*nexusapi.OperationInfo, error) {
```

Comment:
No need for these calls to be on the client (arguably there's not much value in these helpers at all)